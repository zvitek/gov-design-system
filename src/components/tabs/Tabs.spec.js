import { mount } from '@vue/test-utils'
import GovTabs from '@components/tabs/Tabs'
import GovTabItem from '@components/tabs/TabItem'

let wrapper

const WrapperComp = {
    template: `
        <GovTabs value="tab1">
            <GovTabItem value="tab1"></GovTabItem>
            <GovTabItem value="tab2" :visible="false"></GovTabItem>
        </GovTabs>`,
    components: {
        GovTabs, GovTabItem
    }
}

describe('GovTabs', () => {
    beforeEach(() => {
        wrapper = mount(WrapperComp).find(GovTabs)
    })

    it('is called', () => {
        expect(wrapper.name()).toBe('GovTabs')
        expect(wrapper.isVueInstance()).toBeTruthy()
    })

    it('render correctly', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })

    it('emit input event with value when active tab is modified', async () => {
        const idx = 'tab2'
        wrapper.vm.activeId = idx

        await wrapper.vm.$nextTick() // Wait until $emits have been handled

        const valueEmitted = wrapper.emitted().input

        expect(valueEmitted).toBeTruthy()
        expect(valueEmitted[0][0]).toBe(idx)
    })

    it('emit input event with value when childClick is called', async () => {
        const val = 'tab2'

        wrapper.vm.childClick(wrapper.vm.items[1])
        wrapper.vm.childClick(wrapper.vm.items[1])

        await wrapper.vm.$nextTick() // Wait until $emits have been handled

        const valueEmitted = wrapper.emitted().input
        expect(valueEmitted[0][0]).toBe(val)
        // Will be called only once even if we clicked multiple times
        expect(valueEmitted.length).toBe(1)
    })

    it('updates the tab when the value is changed', async () => {
        wrapper.setProps({value: 'tab2'})

        expect(wrapper.vm.activeId).toBe('tab2')

        wrapper.setProps({value: 0})

        expect(wrapper.vm.activeId).toBe('tab1')
    })

    it('finds the children in descendants', () => {
        const testCmp = {
            template: `
                <div>
                    <GovTabItem></GovTabItem>
                    <GovTabItem></GovTabItem>
                </div>`,
            components: {GovTabItem}
        }

        wrapper = mount({
            template: `
                <GovTabs>
                    <test-cmp/>
                    <test-cmp/>
                    <test-cmp/>
                </GovTabs>`,
            components: {
                GovTabs, testCmp
            }
        }).find(GovTabs)

        expect(wrapper.vm.items.length).toBe(6)
        expect(wrapper.vm.items.map((i) => i.index)).toEqual([0, 1, 2, 3, 4, 5])
    })

    it('doesn\'t show a tab if value is null', () => {
        wrapper.setProps({value: null})

        expect(wrapper.vm.activeId).toBeNull()
        expect(wrapper.vm.activeTab).toBe(undefined)
    })

    it('still renders if there is no item', () => {
        wrapper = mount({
            template: `<GovTabs value="tab1"></GovTabs>`,
            components: {
                GovTabs
            }
        }).find(GovTabs)
        expect(wrapper.html()).toMatchSnapshot()
    })
})
