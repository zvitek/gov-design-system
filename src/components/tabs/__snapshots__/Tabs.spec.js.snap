// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GovTabs render correctly 1`] = `
<div class="gov-tabs">
    <nav role="tablist" aria-orientation="horizontal" class="gov-tabs__links-holder"><button role="tab" id="tab1-tab" aria-controls="tab1-content" aria-selected="true" tabindex="0" class="gov-tabs__link is-active">

        </button><button role="tab" id="tab2-tab" aria-controls="tab2-content" aria-selected="false" tabindex="-1" class="gov-tabs__link" style="display: none;">

        </button></nav>
    <div class="gov-tabs__content-holder">
        <div class="gov-tabs__content is-active">
            <div class="gov-tabs__content is-active" role="tabpanel" id="tab1-content" aria-labelledby="tab1-label" tabindex="0"></div>
            <div class="gov-tabs__content is-active" role="tabpanel" id="tab2-content" aria-labelledby="tab2-label" tabindex="-1" style="display: none;"></div>
        </div>
    </div>
</div>
`;

exports[`GovTabs still renders if there is no item 1`] = `
<div class="gov-tabs">
    <nav role="tablist" aria-orientation="horizontal" class="gov-tabs__links-holder"></nav>
    <div class="gov-tabs__content-holder">
        <div class="gov-tabs__content is-active"></div>
    </div>
</div>
`;
